rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,511 @@
+even if the system time changes due to daylight savings time, NTP clock syncs,
+or manually setting the time. To check for expiry, other nodes poll the
+cluster state document. Once they see that the `lastSeen` value has increased,
+they capture the current hr time `current_hr_time` and starts waiting until
+`process.hrtime() - current_hr_time > expiry_timeout` if at that point
+`lastSeen` hasn’t been updated the lease is considered to have expired. This
+means other nodes can take up to `2*expiry_timeout` to recognize an expired
+lease, but a lease will never expire prematurely. 
+
+Any node that detects an expired lease can release that lease by setting the
+expired node’s `hasLease = false`. It can then attempt to acquire its lease. 
+
+## 4.4 Minimize data loss with mixed Kibana versions
+When multiple versions of Kibana are running at the same time, writes from the
+outdated node can end up either in the outdated Kibana index, the newly
+migrated index, or both. New documents added (and some updates) into the old
+index while a migration is in-progress will be lost. Writes that end up in the
+new index will be in an outdated format. This could cause queries on the data
+to only return a subset of the results which leads to incorrect results or
+silent data loss.
+
+Any in-progress writes such as a bulk operation will continue to run even
+after an index was set to read-only or an alias had been removed. There are
+also certain failure scenarios under which ClusterState (such as aliases or
+index settings) remains inconsistent between nodes for an unbounded
+amount of time. Since Kibana cannot rely on Elasticsearch scripting support
+being enabled, we don't have a mechanism to guarantee that there would be no
+data loss. Instead, we aim to minimize it as much as possible.
+
+To minimize data loss we need to implement a “guard” which will prevent old
+nodes and new nodes from writing at the same time. To perform a rollback for a
+failed migration, this guard will have to be removed. By locating this state
+in the backup snapshot taken before an upgrade, rolling back to the previous
+snapshot will automatically remove the guard.
+
+**7.x Minimize data loss when creating new indices**
+
+Preventing data loss from mixed 7.x versions, introduces two additional steps
+to rollback to a previous version without a snapshot:
+1. (existing) Point the `.kibana` alias to the previous Kibana index `.kibana_n-1`
+2. (existing) Delete `.kibana_n`
+3. (new) Enable writes on `.kibana_n-1`
+4. (new) Delete the dummy "version lock" document from `.kibana_n-1`
+
+Since our documentation and server logs have implicitly encouraged users to
+rollback without using snapshots, many users might have to rely on these
+additional migration steps to perform a rollback. Since even the existing
+steps are error prone, introducing more steps will likely introduce more
+problems than what it solves.
```

Comment:
Are you stating that algo below is something we shouldn't do?

Reply:
Yes. I could maybe move this to alternatives as well to make it clearer.

Reply:
I moved the 7.x algorithm to the alternatives section and put the other paragraphs with the 8.x algorithm for preventing data loss with mixed versions.